update code model

iiiidaaa opened the pull request:
新しいEntityに移行するためのmodel,dbとserverのコード修正
- model
CodeGitleaksモデルをCodeGithubSetting,CodeGitleaksSettingに分割
元々CodeGitleaksに存在していたフィールドは以下に分割されます
  - CodeGithubSetting
    - CodeGithubSettingID (CodeGitleaksのGitleaksID)
    - Name
    - ProjectID
    - Type
    - BaseURL
    - TargetResource
    - GithubUser
    - PersonalAccessToken
    - CreatedAt
    - UpdatedAt
  - CodeGitleaksSetting
    - CodeGithubSettingID (CodeGitleaksのGitleaksID、親のCodeGithubSettingと同値にする)
    - CodeDataSourceID
    - ProjectID
    - RepositoryPattern
    - ScanPublic
    - ScanInternal
    - ScanPrivate
    - Status
    - StatusDetail
    - ScanAt
    - CreatedAt
    - UpdatedAt

CodeEnterpriseOrgモデルはCodeGithubEnterpriseOrgに命名変更します
また、フィールドLoginをOrganizationに名前の変更を行います

以下の２パラメータは現状使用されていないため、廃止します(gRPCのレスポンスとしてはゼロ値が返されるようになります)
- GitleaksConfig
- ScanSucceededAt

- DB
modelの修正に合わせて、code_github_setting,code_gitleaks_setting,code_github_enterprise_orgの操作を行うメソッドの追加

- server
Gitleaks関連のServiceは元々1つのテーブルを操作ところが、2つのテーブルを操作するようになったため以下のような挙動へと変更しています。
- ListGitleaks,GetGitleaks
GithubSetting取得後、取得したGithubSettingのIDからGitleaksSettingを取得
2つのデータからレスポンスを作成します
GithubSettingとGitleaksSettingは1:1の関係になるため、取得したGithubSettingのIDを持つGitleaksSettingが存在しない場合エラーを返します
- PutGitleaks
リクエストデータから上記モデルの通りにデータを分割して登録します
- DeleteGitleaks
リクエストデータからCodeGithubSettingを削除
その後、削除したデータを親に持つCodeGitleaksSetting,CodeGithubEnterpriseOrgを削除

- GithubEnterpriseOrg
テーブル(モデル)名の変更とパラメータ名LoginをOrganizationへの変更を行いました

iiiidaaa committed:
update code model

新しいEntityに移行するためのmodel,dbとserverのコード修正
- model
CodeGitleaksモデルをCodeGithubSetting,CodeGitleaksSettingに分割
元々CodeGitleaksに存在していたフィールドは以下に分割されます
  - CodeGithubSetting
    - CodeGithubSettingID (CodeGitleaksのGitleaksID)
    - Name
    - ProjectID
    - Type
    - BaseURL
    - TargetResource
    - GithubUser
    - PersonalAccessToken
    - CreatedAt
    - UpdatedAt
  - CodeGitleaksSetting
    - CodeGithubSettingID (CodeGitleaksのGitleaksID、親のCodeGithubSettingと同値にする)
    - CodeDataSourceID
    - ProjectID
    - RepositoryPattern
    - ScanPublic
    - ScanInternal
    - ScanPrivate
    - Status
    - StatusDetail
    - ScanAt
    - CreatedAt
    - UpdatedAt

CodeEnterpriseOrgモデルはCodeGithubEnterpriseOrgに命名変更します
また、フィールドLoginをOrganizationに名前の変更を行います

以下の２パラメータは現状使用されていないため、廃止します(gRPCのレスポンスとしてはゼロ値が返されるようになります)
- GitleaksConfig
- ScanSucceededAt

- DB
modelの修正に合わせて、code_github_setting,code_gitleaks_setting,code_github_enterprise_orgの操作を行うメソッドの追加

- server
Gitleaks関連のServiceは元々1つのテーブルを操作ところが、2つのテーブルを操作するようになったため以下のような挙動へと変更しています。
- ListGitleaks,GetGitleaks
GithubSetting取得後、取得したGithubSettingのIDからGitleaksSettingを取得
2つのデータからレスポンスを作成します
GithubSettingとGitleaksSettingは1:1の関係になるため、取得したGithubSettingのIDを持つGitleaksSettingが存在しない場合エラーを返します
- PutGitleaks
リクエストデータから上記モデルの通りにデータを分割して登録します
- DeleteGitleaks
リクエストデータからCodeGithubSettingを削除
その後、削除したデータを親に持つCodeGitleaksSetting,CodeGithubEnterpriseOrgを削除

- GithubEnterpriseOrg
テーブル(モデル)名の変更とパラメータ名LoginをOrganizationへの変更を行いました

gassara-kys commented on pkg/server/code/code.go:
```diff
@@ -50,57 +50,54 @@ func (c *CodeService) ListDataSource(ctx context.Context, req *code.ListDataSour
 	data := code.ListGitleaksResponse{}
 	for _, d := range *list {
-		data.Gitleaks = append(data.Gitleaks, convertGitleaks(&d, true))
+		gitleaksSetting, err := c.repository.GetGitleaksSetting(ctx, d.ProjectID, d.CodeGithubSettingID)
```
for文の中にSQL呼び出しがあるので、一回でとってきた方がパフォーマンス的に良さそうだと思いました。

[n+1問題の一般的な対処法]
- JOIN句を使う
- Eager loading

iiiidaaa replied:
どちらもGithubSetting,GitleaksSettingのどちらもListで取ってきて、メモリ上のデータでCodeGithubSettingIDの突き合わせを行うように修正します。(最大2回SQLを叩く)

joinを使って1回のSQLでデータを取ってくることも考えましたが、データ取得用の構造体作成したり、クエリ実行のコードが複雑になりそうだったり、取得したデータの振り分けだったりでコードが多少複雑になりそうなので考慮の外としました。

他何か良さそうな方針などあればご教示いただければと思います。

gassara-kys replied:
> どちらもGithubSetting,GitleaksSettingのどちらもListで取ってきて、メモリ上のデータでCodeGithubSettingIDの突き合わせを行うように修正します。(最大2回SQLを叩く)

ありがとうございます。
上記の場合、パフォーマンス的には解消されそうですが、その反面コードの複雑性は増しそうですね。
データ量が少ない想定で、Repository層はシンプルにしたいということであれば、現状のままでも問題はありません。（その他の箇所での使い回しがききやすいなどのメリットがあるとか）

ただ、各エンティティすべてにCRUDのメソッドを用意するとコード量も多くなりメンテナンス性を懸念していました。（その分テストも書く必要がある）それなら単純なJOINを行うSQL一つで十分かなと思った次第でした。

mmmknt replied:
横からですがコメントです。
Entityとテーブルは1対1じゃなくてもよいと思うので、例えば、code_gitleaks_settingsのプロパティとしてcode_github_settingを持たせて、RISKENからはcode_gitleaks_settingsというEntity単位でしか扱わないようにする選択もあるかと思います。
その場合、構造体を複数用意したり、sqlのクエリやコードが複雑になるといった上で挙げられている課題は解決します。
ただ、この方針だとこれまでとEntityの扱いが変わるため、パフォーマンス上の問題やコードのメンテナンス性の問題がそんなになければ、これまでと同様の方針でも良さそうです。

iiiidaaa replied:
>どちらもGithubSetting,GitleaksSettingのどちらもListで取ってきて、メモリ上のデータでCodeGithubSettingIDの突き合わせを行うように修正します。

この方法で修正しました。

gassara-kys commented on pkg/db/code.go:
```diff
@@ -46,247 +52,241 @@ func (c *Client) ListCodeDataSource(ctx context.Context, codeDataSourceID uint32
+	github_user=VALUES(github_user)
+`
+
+func (c *Client) UpsertGithubSettingWithoutToken(ctx context.Context, data *code.GitleaksForUpsert) (*model.CodeGithubSetting, error) {
```
新規のexpose functionに関しては必ずテスト書くというということだったのでお願いします。

iiiidaaa replied:
code_github_setting,code_gitleaks_setting,code_github_enterprise_orgの操作に関するテストを作成しました。

gassara-kys reviewed (commented):
全体的に `Github` という表記が気になりました。
固有名詞で正式には `GitHub` (Hが大文字)なので、正式名称を使っておく方が良いかな？と思いました。:eyes:

iiiidaaa commented:
@gassara-kys 
>全体的に Github という表記が気になりました。
固有名詞で正式には GitHub (Hが大文字)なので、正式名称を使っておく方が良いと思いました。

protoの修正も考えると、以下のような懸念があったので、hを小文字で統一していました。
message GitHubSetting を作成した際に、以下のような記述になることが想定される
```
message GitHubSetting {
  uint32 code_github_setting_id = 1;
  uint32 code_data_source_id = 2;
...
```
この場合、自動生成される GitHubSettingメッセージのパラメータ名が**CodeGithubSettingId**になる
そのため、model,server上でHを大文字にすると箇所ごとに大文字小文字が混ざってわかりづらくなるかなと考えました。

そのため、いくつか修正案を考えたのですがどの案でいくか相談させてください。
1. protoの自動生成される値のみhの小文字を許容する
2. protoの自動生成される値もhが大文字にする
全ての表記がGitHubに統一されますが、
protoの記述に```uint32 code_git_hub_setting = 1; ```みたいな区切り方をしなければいけないと想定しています。
これだとprotoの記述が変な分け方になってしまってるように感じます。
3. 全てhが小文字になることを許容する

1,2 は今回の範囲であれば全体をhを大文字にする修正を行います。
3は現状のままにする感じです。

個人的には 1. でいいかなと考えていますが、別の修正案などあればご教示いただきたいです。
(そもそも AwsIdのd小文字のように自動生成されるものに関しては既に統一できていない場所もありますし)

gassara-kys commented:
> 個人的には 1. でいいかなと考えていますが、別の修正案などあればご教示いただきたいです。

私も１でいいと思います 🙏

mmmknt commented on pkg/server/code/code.go:
```diff
@@ -133,21 +134,40 @@ func (c *CodeService) PutGitleaks(ctx context.Context, req *code.PutGitleaksRequ
+		return nil, err
+	}
+	req.Gitleaks.GitleaksId = registerdGithubSetting.CodeGithubSettingID
+	registerdGitleaksSetting, err := c.repository.UpsertGitleaksSetting(ctx, req.Gitleaks)
```
typo: registerd -> registered

iiiidaaa replied:
修正しました！

mmmknt commented on pkg/server/code/code.go:
```diff
@@ -133,21 +134,40 @@ func (c *CodeService) PutGitleaks(ctx context.Context, req *code.PutGitleaksRequ
 		return nil, err
 	}
-	err := c.repository.DeleteGitleaks(ctx, req.ProjectId, req.GitleaksId)
+	err := c.repository.DeleteGithubSetting(ctx, req.ProjectId, req.GitleaksId)
```
将来的な話ですが、削除対象のcode_github_settingのデータを他のテーブルが参照していないかをチェックして、参照していたらこのAPIでは削除しないようにする必要があると思います。
このPRで対応が必要というわけではないですが、覚えておいてもらえると:pray:

iiiidaaa replied:
ありがとうございます！承知しました！

mmmknt commented on pkg/server/code/code.go:
```diff
@@ -244,41 +264,29 @@ func (c *CodeService) InvokeScanGitleaks(ctx context.Context, req *code.InvokeSc
 	}
 	resp, err := c.sqs.Send(ctx, c.sqs.CodeGitleaksQueueURL, &message.GitleaksQueueMessage{
-		GitleaksID: data.GitleaksID,
+		GitleaksID: data.CodeGithubSettingID,
```
GitleaksIDかGithubSettingIDのどちらかに統一した方が分かりやすくなると思います。

iiiidaaa replied:
ご指摘ありがとうございます！
メッセージのパラメータ名を修正するご指摘だと認識していますが、パラメータ名の変更はGitleaks側への影響も発生することが想定されるためこのPRでは修正せず、以下の流れで修正させてください。

1. 本PRではメッセージのパラメータを変更しない
2. 本PRマージ後、旧テーブルから新テーブルへのデータ移行
3. データ移行後、新しいEntityに合わせたproto,serviceの修正(合わせて、gitleaks,gatewayの修正)

3のタイミングでproto,serviceからGitleaksIDというパラメータがなくなるため、この際に修正したいと思います。

mmmknt replied:
その流れでいいと思います 👍

iiiidaaa committed:
fix typo

iiiidaaa committed:
fix model Github

Githubと記述していた内容をGitHub(h大文字に切り替えます)
Modelに関して以下の追加を行いました
- struct CodeGitHubSettingはcode_git_hub_settingと認識されるためTablerインターフェイスを実装
- GitHub~~パラメータも同様にgit_hubになるためタグを付与

iiiidaaa committed:
fix model db

CodeGitHubEnterpriseOrgのTablerインターフェイスを実装
listGitleaksSettingのSQL分の誤りを修正

iiiidaaa committed:
fix listGitleaks

GitleaksSettingをListで取得することで、SQLの呼び出し回数を減らします
listで取得した、GitHubSettingとGitleaksSettingから結果を生成します

iiiidaaa committed:
add code db test

iiiidaaa commented:
>全体的に Github という表記が気になりました。
固有名詞で正式には GitHub (Hが大文字)なので、正式名称を使っておく方が良いと思いました。

全体的にstruct名、変数名にGitHub(大文字H)を使用するように変更しました。
Modelのstruct,パラメータ名には追加で以下の対応をしています。
Model Struct名
`CodeGitHubSetting`のようにHを大文字にすると、gormがテーブル名を`code_git_hub_setting`と認識するため、以下のようにテーブル名を明示しています。(CodeGitHubEnterpriseOrgも同様)
```
func (CodeGitHubSetting) TableName() string {
	return "code_github_setting"
}
```
Model パラメータ名
`CodeGitHubSettingID`は、`code_git_hub_setting_id`のように解釈されるため、`gorm:"code_github_setting_id"`のようにタグをつけています。(Modelのパラメータ名におけるその他のH大文字も同様)

gassara-kys commented on pkg/server/code/code.go:
```diff
@@ -50,57 +50,62 @@ func (c *CodeService) ListDataSource(ctx context.Context, req *code.ListDataSour
+		return &data, nil
+	}
+	listGitleaksSetting, err := c.repository.ListGitleaksSetting(ctx, req.ProjectId)
+	if err != nil {
```
DBのリレーション的にRecordNotFoundが発生する可能性がありますか？（errを返して問題ないですか？）

iiiidaaa replied:
スライスを用いて、データベースにクエリを投げているのでRecordNotFoundエラーが発生する可能性はないと判断しています。
https://pkg.go.dev/github.com/jinzhu/gorm#pkg-variables
```
// ErrRecordNotFound returns a "record not found error". Occurs only when attempting to query the database with a struct; querying with a slice won't return this error
	ErrRecordNotFound = errors.New("record not found")
	
```
こちらにも単一のレコードを探す際のみにErrRecordNotFoundが発生すると記載あります。
https://gorm.io/ja_JP/docs/query.html

gassara-kys replied:
すいません、ありがとうございます。理解しました！

gassara-kys reviewed (approved):
LGTM

mmmknt commented on pkg/server/code/code.go:
```diff
@@ -50,57 +50,62 @@ func (c *CodeService) ListDataSource(ctx context.Context, req *code.ListDataSour
+	if err != nil {
+		return nil, err
+	}
+	for _, githubSetting := range *listGitHubSetting {
```
データ量を考えると性能上の問題はないかもしれませんが、事前処理で片方のスライスからCodeGithubSettingIDをkeyにしたmapを作っておき、もう片方のスライスをループで回しながらそのmapからデータを取得するようにすると、計算量が一桁落とせます。
メモリ上でのO(n^2)なので対応は必須ではないです。

mmmknt reviewed (approved):
👍
